File: ts/src/test/Exchange/base/test.market.ts
import assert from 'assert';
import Precise from '../../../base/Precise.js';
import { Exchange, Market } from "../../../../ccxt";
import testSharedMethods from './test.sharedMethods.js';

function testMarket (exchange: Exchange, skippedProperties: object, method: string, market: Market) {
    const format = {
        'id': 'btcusd', // string literal for referencing within an exchange
        'symbol': 'BTC/USD', // uppercase string literal of a pair of currencies
        'base': 'BTC', // unified uppercase string, base currency, 3 or more letters
        'quote': 'USD', // unified uppercase string, quote currency, 3 or more letters
        'taker': exchange.parseNumber ('0.0011'), // taker fee, for example, 0.0011 = 0.11%
        'maker': exchange.parseNumber ('0.0009'), // maker fee, for example, 0.0009 = 0.09%
        'baseId': 'btc', // exchange-specific base currency id
        'quoteId': 'usd', // exchange-specific quote currency id
        'active': false, // boolean, market status
        'type': 'spot',
        'linear': false,
        'inverse': false,
        'spot': false,
        'swap': false,
        'future': false,
        'option': false,
        'margin': false,
        'contract': false,
        'contractSize': exchange.parseNumber ('0.001'),
        'expiry': 1656057600000,
        'expiryDatetime': '2022-06-24T08:00:00.000Z',
        'optionType': 'put',
        'strike': exchange.parseNumber ('56000'),
        'settle': 'XYZ',
        'settleId': 'Xyz',
        'precision': {
            // todo : handle precision types after another PR is merged
            'price': exchange.parseNumber ('0.001'), // integer or fraction
            'amount': exchange.parseNumber ('0.001'), // integer or fraction
            'cost': exchange.parseNumber ('0.001'), // integer or fraction
        },
        // value limits when placing orders on this market
        'limits': {
            'amount': {
                'min': exchange.parseNumber ('0.01'), // order amount should be > min
                'max': exchange.parseNumber ('1000'), // order amount should be < max
            },
            'price': {
                'min': exchange.parseNumber ('0.01'), // order price should be > min
                'max': exchange.parseNumber ('1000'), // order price should be < max
            },
            // order cost = price * amount
            'cost': {
                'min': exchange.parseNumber ('0.01'), // order cost should be > min
                'max': exchange.parseNumber ('1000'), // order cost should be < max
            },
        },
        'info': {},
    };
    const emptyAllowedFor = [ 'linear', 'inverse', 'settle', 'settleId', 'expiry', 'expiryDatetime', 'optionType', 'strike', 'margin', 'contractSize' ];
    testSharedMethods.assertStructure (exchange, skippedProperties, method, market, format, emptyAllowedFor);
    testSharedMethods.assertSymbol (exchange, skippedProperties, method, market, 'symbol');
    const logText = testSharedMethods.logTemplate (exchange, method, market);
    //
    const validTypes = [ 'spot', 'margin', 'swap', 'future', 'option', 'index' ];
    testSharedMethods.assertInArray (exchange, skippedProperties, method, market, 'type', validTypes);
    const hasIndex = ('index' in market); // todo: add in all
    // check if string is consistent with 'type'
    if (market['spot']) {
        assert (market['type'] === 'spot', '"type" string should be "spot" when spot is true' + logText);
    } else if (market['swap']) {
        assert (market['type'] === 'swap', '"type" string should be "swap" when swap is true' + logText);
    } else if (market['future']) {
        assert (market['type'] === 'future', '"type" string should be "future" when future is true' + logText);
    } else if (market['option']) {
        assert (market['type'] === 'option', '"type" string should be "option" when option is true' + logText);
    } else if (hasIndex && market['index']) {
        // todo: add index in all implementations
        assert (market['type'] === 'index', '"type" string should be "index" when index is true' + logText);
    }
    // margin check (todo: add margin as mandatory, instead of undefined)
    if (market['spot']) {
        // for spot market, 'margin' can be either true/false or undefined
        testSharedMethods.assertInArray (exchange, skippedProperties, method, market, 'margin', [ true, false, undefined ]);
    } else {
        // otherwise, it must be false or undefined
        testSharedMethods.assertInArray (exchange, skippedProperties, method, market, 'margin', [ false, undefined ]);
    }
    if (!('contractSize' in skippedProperties)) {
        if (!market['spot']) {
            // if not spot, then contractSize should be defined
            assert (market['contractSize'] !== undefined, '"contractSize" must be defined when "spot" is false' + logText);
        }
        testSharedMethods.assertGreater (exchange, skippedProperties, method, market, 'contractSize', '0');
    }
    // typical values
    testSharedMethods.assertGreater (exchange, skippedProperties, method, market, 'expiry', '0');
    testSharedMethods.assertGreater (exchange, skippedProperties, method, market, 'strike', '0');
    testSharedMethods.assertInArray (exchange, skippedProperties, method, market, 'optionType', [ 'put', 'call' ]);
    testSharedMethods.assertGreater (exchange, skippedProperties, method, market, 'taker', '-100');
    testSharedMethods.assertGreater (exchange, skippedProperties, method, market, 'maker', '-100');
    // 'contract' boolean check
    if (market['future'] || market['swap'] || market['option'] || (hasIndex && market['index'])) {
        // if it's some kind of contract market, then `conctract` should be true
        assert (market['contract'], '"contract" must be true when "future", "swap", "option" or "index" is true' + logText);
    } else {
        assert (!market['contract'], '"contract" must be false when neither "future", "swap","option" or "index" is true' + logText);
    }
    const isSwapOrFuture = market['swap'] || market['future'];
    const contractSize = exchange.safeString (market, 'contractSize');
    // contract fields
    if (market['contract']) {
        // linear & inverse should have different values (true/false)
        // todo: expand logic on other market types
        if (isSwapOrFuture) {
            assert (market['linear'] !== market['inverse'], 'market linear and inverse must not be the same' + logText);
            if (!('contractSize' in skippedProperties)) {
                // contract size should be defined
                assert (contractSize !== undefined, '"contractSize" must be defined when "contract" is true' + logText);
                // contract size should be above zero
                assert (Precise.stringGt (contractSize, '0'), '"contractSize" must be > 0 when "contract" is true' + logText);
            }
            if (!('settle' in skippedProperties)) {
                // settle should be defined
                assert ((market['settle'] !== undefined) && (market['settleId'] !== undefined), '"settle" & "settleId" must be defined when "contract" is true' + logText);
            }
        }
        // spot should be false
        assert (!market['spot'], '"spot" must be false when "contract" is true' + logText);
    } else {
        // linear & inverse needs to be undefined
        assert ((market['linear'] === undefined) && (market['inverse'] === undefined), 'market linear and inverse must be undefined when "contract" is false' + logText);
        // contract size should be undefined
        if (!('contractSize' in skippedProperties)) {
            assert (contractSize === undefined, '"contractSize" must be undefined when "contract" is false' + logText);
        }
        // settle should be undefined
        assert ((market['settle'] === undefined) && (market['settleId'] === undefined), '"settle" must be undefined when "contract" is false' + logText);
        // spot should be true
        assert (market['spot'], '"spot" must be true when "contract" is false' + logText);
    }
    // option fields
    if (market['option']) {
        // if option, then strike and optionType should be defined
        assert (market['strike'] !== undefined, '"strike" must be defined when "option" is true' + logText);
        assert (market['optionType'] !== undefined, '"optionType" must be defined when "option" is true' + logText);
    } else {
        // if not option, then strike and optionType should be undefined
        assert (market['strike'] === undefined, '"strike" must be undefined when "option" is false' + logText);
        assert (market['optionType'] === undefined, '"optionType" must be undefined when "option" is false' + logText);
    }
    // future, swap and option should be mutually exclusive
    if (market['future']) {
        assert (!market['swap'] && !market['option'], 'market swap and option must be false when "future" is true' + logText);
    } else if (market['swap']) {
        assert (!market['future'] && !market['option'], 'market future and option must be false when "swap" is true' + logText);
    } else if (market['option']) {
        assert (!market['future'] && !market['swap'], 'market future and swap must be false when "option" is true' + logText);
    }
    // expiry field
    if (market['future'] || market['option']) {
        // future or option markets need 'expiry' and 'expiryDatetime'
        assert (market['expiry'] !== undefined, '"expiry" must be defined when "future" is true' + logText);
        assert (market['expiryDatetime'] !== undefined, '"expiryDatetime" must be defined when "future" is true' + logText);
        // expiry datetime should be correct
        const isoString = exchange.iso8601 (market['expiry']);
        assert (market['expiryDatetime'] === isoString, 'expiryDatetime ("' + market['expiryDatetime'] + '") must be equal to expiry in iso8601 format "' + isoString + '"' + logText);
    } else {
        // otherwise, they need to be undefined
        assert ((market['expiry'] === undefined) && (market['expiryDatetime'] === undefined), '"expiry" and "expiryDatetime" must be undefined when it is not future|option market' + logText);
    }
    // check precisions
    if (!('precision' in skippedProperties)) {
        const precisionKeys = Object.keys (market['precision']);
        const keysLength = precisionKeys.length;
        assert (keysLength >= 2, 'precision should have "amount" and "price" keys at least' + logText);
        for (let i = 0; i < precisionKeys.length; i++) {
            testSharedMethods.checkPrecisionAccuracy (exchange, skippedProperties, method, market['precision'], precisionKeys[i]);
        }
    }
    // check limits
    if (!('limits' in skippedProperties)) {
        const limitsKeys = Object.keys (market['limits']);
        const keysLength = limitsKeys.length;
        assert (keysLength >= 3, 'limits should have "amount", "price" and "cost" keys at least' + logText);
        for (let i = 0; i < limitsKeys.length; i++) {
            const key = limitsKeys[i];
            const limitEntry = market['limits'][key];
            // min >= 0
            testSharedMethods.assertGreaterOrEqual (exchange, skippedProperties, method, limitEntry, 'min', '0');
            // max >= 0
            testSharedMethods.assertGreater (exchange, skippedProperties, method, limitEntry, 'max', '0');
            // max >= min
            const minString = exchange.safeString (limitEntry, 'min');
            if (minString !== undefined) {
                testSharedMethods.assertGreaterOrEqual (exchange, skippedProperties, method, limitEntry, 'max', minString);
            }
        }
    }
    // check whether valid currency ID and CODE is used
    if (!('currency' in skippedProperties) && !('currencyIdAndCode' in skippedProperties)) {
        testSharedMethods.assertValidCurrencyIdAndCode (exchange, skippedProperties, method, market, market['baseId'], market['base']);
        testSharedMethods.assertValidCurrencyIdAndCode (exchange, skippedProperties, method, market, market['quoteId'], market['quote']);
        testSharedMethods.assertValidCurrencyIdAndCode (exchange, skippedProperties, method, market, market['settleId'], market['settle']);
    }
    testSharedMethods.assertTimestamp (exchange, skippedProperties, method, market, undefined, 'created');
}

export default testMarket;

File: julia/Ccxt/test/validators/test.market.jl
using Test
using Ccxt
function testMarket(exchange, skippedProperties, method, market)

    if functions.ccxtruthy(market == nothing)
            return 
    end
    format = Dict{Symbol, Any}(
        Symbol("id") => "btcusd",
        Symbol("symbol") => "BTC/USD",
        Symbol("base") => "BTC",
        Symbol("quote") => "USD",
        Symbol("taker") => parseNumber(exchange, "0.0011"),
        Symbol("maker") => parseNumber(exchange, "0.0009"),
        Symbol("baseId") => "btc",
        Symbol("quoteId") => "usd",
        Symbol("active") => false,
        Symbol("type") => "spot",
        Symbol("linear") => false,
        Symbol("inverse") => false,
        Symbol("spot") => false,
        Symbol("swap") => false,
        Symbol("future") => false,
        Symbol("option") => false,
        Symbol("margin") => false,
        Symbol("contract") => false,
        Symbol("contractSize") => parseNumber(exchange, "0.001"),
        Symbol("expiry") => 1656057600000,
        Symbol("expiryDatetime") => "2022-06-24T08:00:00.000Z",
        Symbol("optionType") => "put",
        Symbol("strike") => parseNumber(exchange, "56000"),
        Symbol("settle") => "XYZ",
        Symbol("settleId") => "Xyz",
        Symbol("precision") => Dict{Symbol, Any}(
            Symbol("price") => parseNumber(exchange, "0.001"),
            Symbol("amount") => parseNumber(exchange, "0.001"),
            Symbol("cost") => parseNumber(exchange, "0.001")
        ),
        Symbol("limits") => Dict{Symbol, Any}(
            Symbol("amount") => Dict{Symbol, Any}(
                Symbol("min") => parseNumber(exchange, "0.01"),
                Symbol("max") => parseNumber(exchange, "1000")
            ),
            Symbol("price") => Dict{Symbol, Any}(
                Symbol("min") => parseNumber(exchange, "0.01"),
                Symbol("max") => parseNumber(exchange, "1000")
            ),
            Symbol("cost") => Dict{Symbol, Any}(
                Symbol("min") => parseNumber(exchange, "0.01"),
                Symbol("max") => parseNumber(exchange, "1000")
            )
        ),
        Symbol("marginModes") => Dict{Symbol, Any}(
            Symbol("cross") => true,
            Symbol("isolated") => false
        ),
        Symbol("info") => Dict{Symbol, Any}()
    );
    if functions.ccxtruthy(ccxt_in("quanto", market))
        format[Symbol("quanto")] = false;
    end
    spot = get(market, Symbol("spot"), nothing);
    contract = get(market, Symbol("contract"), nothing);
    swap = get(market, Symbol("swap"), nothing);
    future = get(market, Symbol("future"), nothing);
    option = get(market, Symbol("option"), nothing);
    index = safeBool(exchange, market, "index");
    isIndex = @functions.ccxt_and((index != nothing), index);
    linear = get(market, Symbol("linear"), nothing);
    inverse = get(market, Symbol("inverse"), nothing);
    quanto = safeBool(exchange, market, "quanto");
    isQuanto = @functions.ccxt_and((quanto != nothing), quanto);
    isInactiveMarket = get(market, Symbol("active"), nothing) == false;
    emptyAllowedFor = ["margin"];
    if functions.ccxtruthy(!functions.ccxtruthy(contract))
                push!(emptyAllowedFor, "contractSize");
                push!(emptyAllowedFor, "linear");
                push!(emptyAllowedFor, "inverse");
                push!(emptyAllowedFor, "quanto");
                push!(emptyAllowedFor, "settle");
                push!(emptyAllowedFor, "settleId");
    end
    if functions.ccxtruthy(@functions.ccxt_and(!functions.ccxtruthy(future), !functions.ccxtruthy(option)))
                push!(emptyAllowedFor, "expiry");
                push!(emptyAllowedFor, "expiryDatetime");
    end
    if functions.ccxtruthy(!functions.ccxtruthy(option))
                push!(emptyAllowedFor, "optionType");
                push!(emptyAllowedFor, "strike");
    end
    if functions.ccxtruthy(isInactiveMarket)
                push!(emptyAllowedFor, "contractSize");
                push!(emptyAllowedFor, "settle");
                push!(emptyAllowedFor, "settleId");
                push!(emptyAllowedFor, "baseId");
                push!(emptyAllowedFor, "quoteId");
                push!(emptyAllowedFor, "base");
                push!(emptyAllowedFor, "quote");
    end
    assertStructure(testSharedMethods, exchange, skippedProperties, method, market, format, emptyAllowedFor);
    assertSymbol(testSharedMethods, exchange, skippedProperties, method, market, "symbol");
    logText = logTemplate(testSharedMethods, exchange, method, market);
    assertGreater(testSharedMethods, exchange, skippedProperties, method, market, "taker", "-100");
    assertLess(testSharedMethods, exchange, skippedProperties, method, market, "taker", "100");
    assertGreater(testSharedMethods, exchange, skippedProperties, method, market, "maker", "-100");
    assertLess(testSharedMethods, exchange, skippedProperties, method, market, "maker", "100");
    validTypes = ["spot", "margin", "swap", "future", "option", "index", "other"];
    assertInArray(testSharedMethods, exchange, skippedProperties, method, market, "type", validTypes);
    validSubTypes = ["linear", "inverse", "quanto", nothing];
    assertInArray(testSharedMethods, exchange, skippedProperties, method, market, "subType", validSubTypes);
    checkedTypes = ["spot", "swap", "future", "option"];
    i = 0
    while functions.ccxtruthy(functions.ccxt_lt(i, length(checkedTypes)))
        type_var = get(checkedTypes, i + 1, nothing);
        if functions.ccxtruthy(get(market, Symbol(type_var), nothing))
            @test type_var == get(market, Symbol("type"), nothing)
        end
        i += 1
    end
    if functions.ccxtruthy(@functions.ccxt_or(swap, future))
        checkedSubTypes = ["linear", "inverse"];
        i = 0
        while functions.ccxtruthy(functions.ccxt_lt(i, length(checkedSubTypes)))
            subType = get(checkedSubTypes, i + 1, nothing);
            if functions.ccxtruthy(get(market, Symbol(subType), nothing))
                @test subType == get(market, Symbol("subType"), nothing)
            end
            i += 1
        end

    end
    if functions.ccxtruthy(spot)
        assertInArray(testSharedMethods, exchange, skippedProperties, method, market, "margin", [true, false, nothing]);
    else
        assertInArray(testSharedMethods, exchange, skippedProperties, method, market, "margin", [false, nothing]);
    end
    if functions.ccxtruthy(spot)
        @test functions.ccxtruthy(@functions.ccxt_and(@functions.ccxt_and(@functions.ccxt_and(@functions.ccxt_and(@functions.ccxt_and(!functions.ccxtruthy(contract), linear == nothing), inverse == nothing), !functions.ccxtruthy(option)), !functions.ccxtruthy(swap)), !functions.ccxtruthy(future)))
    else
        @test functions.ccxtruthy(@functions.ccxt_and(contract, (@functions.ccxt_or(@functions.ccxt_or(@functions.ccxt_or(future, swap), option), isIndex))))
    end
    contractSize = safeString(exchange, market, "contractSize");
    if functions.ccxtruthy(@functions.ccxt_and(contract, !functions.ccxtruthy(isInactiveMarket)))
        if functions.ccxtruthy(isQuanto)
            @test linear == false
            @test inverse == false
        else
            @test inverse != nothing
            @test linear != nothing
            @test linear != inverse
        end
        @test functions.ccxtruthy((@functions.ccxt_or((ccxt_in("contractSize", skippedProperties)), contractSize != nothing)))
        @test functions.ccxtruthy(@functions.ccxt_or((ccxt_in("contractSize", skippedProperties)), stringGt(contractSize, "0")))
        @test functions.ccxtruthy(@functions.ccxt_or((ccxt_in("settle", skippedProperties)), (@functions.ccxt_and(get(market, Symbol("settle"), nothing) != nothing, get(market, Symbol("settleId"), nothing) != nothing))))
    elseif functions.ccxtruthy(!functions.ccxtruthy(contract))
        @test functions.ccxtruthy(@functions.ccxt_and(@functions.ccxt_and(linear == nothing, inverse == nothing), quanto == nothing))
        @test contractSize == nothing
        @test functions.ccxtruthy(@functions.ccxt_and((get(market, Symbol("settle"), nothing) == nothing), (get(market, Symbol("settleId"), nothing) == nothing)))
    end
    if functions.ccxtruthy(get(market, Symbol("future"), nothing))
        @test functions.ccxtruthy(@functions.ccxt_and(@functions.ccxt_and(!functions.ccxtruthy(get(market, Symbol("swap"), nothing)), !functions.ccxtruthy(get(market, Symbol("option"), nothing))), !functions.ccxtruthy(isIndex)))
    elseif functions.ccxtruthy(get(market, Symbol("swap"), nothing))
        @test functions.ccxtruthy(@functions.ccxt_and(!functions.ccxtruthy(get(market, Symbol("future"), nothing)), !functions.ccxtruthy(get(market, Symbol("option"), nothing))))
    else
        if functions.ccxtruthy(get(market, Symbol("option"), nothing))
            @test functions.ccxtruthy(@functions.ccxt_and(!functions.ccxtruthy(get(market, Symbol("future"), nothing)), !functions.ccxtruthy(get(market, Symbol("swap"), nothing))))
        end

    end
    if functions.ccxtruthy(@functions.ccxt_or(option, future))
        @test get(market, Symbol("expiry"), nothing) != nothing
        @test get(market, Symbol("expiryDatetime"), nothing) != nothing
        isoString = iso8601(exchange, get(market, Symbol("expiry"), nothing));
        @test get(market, Symbol("expiryDatetime"), nothing) == isoString
        assertGreater(testSharedMethods, exchange, skippedProperties, method, market, "expiry", "0");
        if functions.ccxtruthy(option)
            @test functions.ccxtruthy((@functions.ccxt_or((ccxt_in("strike", skippedProperties)), get(market, Symbol("strike"), nothing) != nothing)))
            assertGreater(testSharedMethods, exchange, skippedProperties, method, market, "strike", "0");
            @test functions.ccxtruthy((@functions.ccxt_or((ccxt_in("optionType", skippedProperties)), get(market, Symbol("optionType"), nothing) != nothing)))
            assertInArray(testSharedMethods, exchange, skippedProperties, method, market, "optionType", ["put", "call"]);
        else
            @test get(market, Symbol("strike"), nothing) == nothing
            @test get(market, Symbol("optionType"), nothing) == nothing
        end
    elseif functions.ccxtruthy(spot)
        @test functions.ccxtruthy(@functions.ccxt_and((get(market, Symbol("expiry"), nothing) == nothing), (get(market, Symbol("expiryDatetime"), nothing) == nothing)))
    end
    precisionKeys = objectKeys(get(market, Symbol("precision"), nothing));
    precisionKeysLen = length(precisionKeys);
    @test functions.ccxtruthy(functions.ccxt_ge(precisionKeysLen, 2))
    i = 0
    while functions.ccxtruthy(functions.ccxt_lt(i, length(precisionKeys)))
        priceOrAmountKey = get(precisionKeys, i + 1, nothing);
        isExclusivePair = get(market, Symbol("baseId"), nothing) == "BTC";
        isNonSpot = !functions.ccxtruthy(spot);
        isPrice = priceOrAmountKey == "price";
        isTickSize5 = stringEq("5", safeString(exchange, get(market, Symbol("precision"), nothing), priceOrAmountKey));
        if functions.ccxtruthy(@functions.ccxt_and(@functions.ccxt_and(@functions.ccxt_and(isNonSpot, isPrice), isExclusivePair), isTickSize5))
            i += 1; continue
        end
        if functions.ccxtruthy(!functions.ccxtruthy((ccxt_in("precision", skippedProperties))))
            checkPrecisionAccuracy(testSharedMethods, exchange, skippedProperties, method, get(market, Symbol("precision"), nothing), priceOrAmountKey);
        end
        i += 1
    end
    limitsKeys = objectKeys(get(market, Symbol("limits"), nothing));
    limitsKeysLength = length(limitsKeys);
    @test functions.ccxtruthy(functions.ccxt_ge(limitsKeysLength, 3))
    i = 0
    while functions.ccxtruthy(functions.ccxt_lt(i, length(limitsKeys)))
        key = get(limitsKeys, i + 1, nothing);
        limitEntry = get(get(market, Symbol("limits"), nothing), Symbol(key), nothing);
        if functions.ccxtruthy(isInactiveMarket)
            i += 1; continue
        end
        if functions.ccxtruthy(!functions.ccxtruthy((ccxt_in("limits", skippedProperties))))
            assertGreaterOrEqual(testSharedMethods, exchange, skippedProperties, method, limitEntry, "min", "0");
            assertGreater(testSharedMethods, exchange, skippedProperties, method, limitEntry, "max", "0");
            minString = safeString(exchange, limitEntry, "min");
            if functions.ccxtruthy(minString != nothing)
                assertGreaterOrEqual(testSharedMethods, exchange, skippedProperties, method, limitEntry, "max", minString);
            end
        end
        i += 1
    end
    if functions.ccxtruthy(!functions.ccxtruthy(isInactiveMarket))
        assertValidCurrencyIdAndCode(testSharedMethods, exchange, skippedProperties, method, market, get(market, Symbol("baseId"), nothing), get(market, Symbol("base"), nothing));
        assertValidCurrencyIdAndCode(testSharedMethods, exchange, skippedProperties, method, market, get(market, Symbol("quoteId"), nothing), get(market, Symbol("quote"), nothing));
        assertValidCurrencyIdAndCode(testSharedMethods, exchange, skippedProperties, method, market, get(market, Symbol("settleId"), nothing), get(market, Symbol("settle"), nothing));
    end
    assertTimestamp(testSharedMethods, exchange, skippedProperties, method, market, nothing, "created");
    if functions.ccxtruthy(!functions.ccxtruthy((ccxt_in("marginModes", skippedProperties))))
        marginModes = safeDict(exchange, market, "marginModes", Dict{Symbol, Any}());
        @test functions.ccxtruthy(ccxt_in("cross", marginModes))
        @test functions.ccxtruthy(ccxt_in("isolated", marginModes))
        assertInArray(testSharedMethods, exchange, skippedProperties, method, marginModes, "cross", [true, false, nothing]);
        assertInArray(testSharedMethods, exchange, skippedProperties, method, marginModes, "isolated", [true, false, nothing]);
    end
end
